root = '/scratch/genevol/users/lucas/'
rootSave = '/raid/genevol/users/lucas/heritability/00.Descriptive/Images/Hardy/'

##################################################################################################################################
# Compilation of MAF and HWE calculations by snp and chromosome generated by previous programs
# Based on different cutpoints, the objective is to generate lists of snps to enter in posterior pruning analysis - VIF
##################################################################################################################################

hweDf = NULL
mafDf = NULL
refDf = NULL

# Reading HWE files and SNP's mappings
for(i in 1:22){

    print(paste0('Chromosome: ',i))
    # Reading p-value from hwe test for each SNP from i-th chromosome
    # Hwe test from individuals whose phenotype is known, considering only europeans (n=358)
    auxNafrHwe = data.table::fread(paste0(root,'Hardy_NAfr_chr',i,'.hwe'), header = T,select = c('CHR' , 'SNP' , 'P')) %>% 
    mutate(Sample = 'NAfr') # Labelling df
    # Maf from individuals whose phenotype is known, considering only europeans (n=358)
    auxNafrMaf = data.table::fread(paste0(root,'Maf/Hardy_NAfr_chr',i,'.frq'), header = T,select = c('CHR' , 'SNP' , 'MAF')) %>% 
    mutate(Sample = 'NAfr') # Labelling df
    # Hwe test from individuals whose phenotype is known (n=445)
    auxGeuvadisHwe = data.table::fread(paste0(root,'Hardy_Geuvadis_chr',i,'.hwe'), header = T,select = c('CHR' , 'SNP' , 'P')) %>% 
    mutate(Sample = 'Geuvadis') # Labelling df
    # Maf from individuals whose phenotype is known (n=445)
    auxGeuvadisMaf = data.table::fread(paste0(root,'Maf/Hardy_Geuvadis_chr',i,'.frq'), header = T,select = c('CHR' , 'SNP' , 'MAF')) %>% 
    mutate(Sample = 'Geuvadis') # Labelling df
    # Hwe test from all genotyped individuals (n=2504)
    auxFullHwe = data.table::fread(paste0(root,'Hardy_Full_chr',i,'.hwe'), header = T,select = c('CHR' , 'SNP' , 'P')) %>% 
    mutate(Sample = 'Full') # Labelling df
    # Maf from all genotyped individuals (n=2504)
    auxFullMaf = data.table::fread(paste0(root,'Maf/Hardy_Full_chr',i,'.frq'), header = T,select = c('CHR' , 'SNP' , 'MAF')) %>% 
    mutate(Sample = 'Full') # Labelling df
    # Reading SNP's position
    auxRef = data.table::fread(paste0(root,'basicInfoChr',i,'.txt'), header = F)
    colnames(auxRef) = c('CHR','SNP','BP') 
    
    # Appending tests
    hweDf = rbind(hweDf , rbind(auxNafrHwe,rbind(auxGeuvadisHwe,auxFullHwe)))
    # Appending MAF
    mafDf = rbind(mafDf , rbind(auxNafrMaf,rbind(auxGeuvadisMaf,auxFullMaf)))
    # Appending mapping from SNPs
    refDf = rbind(refDf , auxRef)

}

# merge between hwe test and SNP positions
finalPltDfHwe = merge(hweDf,refDf)
# merge between MAF and SNP positions
finalPltDf = merge(finalPltDfHwe,mafDf , by = c("CHR","SNP","Sample") , all.x = T , all.y = T)

# Cutpoints and filters 
samples_ = c('NAfr','Geuvadis','Full')
hwes_ = c(.05 , 1e-2 , 1e-3 , 1e-4 , 1e-5 , 1e-6 , 1e-7 , 1e-8 )
mafs_ = c(.001 , .005 , (1:5)/100)

# Generating list of snps 
cont = 0
for(sample_ in samples_){
  for(maf_ in mafs_){
    for(hwe_ in hwes_){
      aux = finalPltDf %>% filter(Sample == sample_ , P >= hwe_ , MAF >= maf_)
      saveTabl = aux %>% distinct(SNP)
      print(paste(sample_,maf_,hwe_,nrow(aux),cont))
      write.table(saveTabl[,c('SNP')] , paste0(root,'ListSnpFilters/filter_sample_',sample_,'_maf_',maf_,'_hwe_',hwe_,'.txt'), col.names = F , row.names = F,quote = F)
      cont = cont + 1
    }
  } 
}

##################################################################################################################################
# Relation between volume of snps and cutpoints previous to pruning analysis
##################################################################################################################################

suppressMessages(suppressWarnings(library(dplyr)))
suppressMessages(suppressWarnings(library(ggplot2)))
suppressMessages(suppressWarnings(library(gtable)))

# Defining important plot functions (Manhattan plt)
manhattanPltDf = function(df_,samp_){
#adapted from:https://www.r-graph-gallery.com/101_Manhattan_plot.html
    dfPrep <- df_ %>% 
  # Compute chromosome size
  group_by(CHR) %>% 
  summarise(chr_len=as.numeric(max(BP))) %>% 
  # Calculate cumulative position of each chromosome
  mutate(tot=cumsum(chr_len)-chr_len) %>%
  select(-chr_len) %>%
  # Add this info to the initial dataset
  left_join(df_, ., by=c("CHR"="CHR")) %>%
  # Add a cumulative position of each SNP
  arrange(CHR, BP) %>%
  mutate(BPcum=BP+tot , Samp = samp_)

  axisdf = dfPrep %>% group_by(CHR) %>% summarize(center=( max(BPcum) + min(BPcum) ) / 2 )%>%
  ungroup() %>%
  mutate(Samp = samp_)

  return(list(dfPrep,axisdf))

}

manhattanPlt = function(dfPrep_,axisdf_,cutPoint_,type_){
#adapted from:https://www.r-graph-gallery.com/101_Manhattan_plot.html

  if(type_ == 'hwe'){
    plt0 = ggplot(dfPrep_, aes(x=BPcum, y=-log10(P)))
  } else{
    plt0 = ggplot(dfPrep_, aes(x=BPcum, y=MAF))
  }
    plt = plt0 +
    # Show all points
    geom_point( aes(color=as.factor(CHR)), alpha=0.8, size=1.3) +
    
    # custom X axis:
    scale_x_continuous( label = axisdf_$CHR, breaks= axisdf_$center ) +
    scale_y_continuous(expand = c(0, 0) ) +     # remove space between plot area and x axis

    # Custom the theme:
    theme_bw() +
    theme( 
      legend.position="none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank()
    ) +
    geom_hline(yintercept = cutPoint_ , col = 'red') +
    facet_wrap(~Samp , ncol = 1)
    
    return(plt)
}

savePlt = function(df_,sample_,root_,name_){

    dfPlot = df_ %>% filter(Sample == sample_)
    p = manhattanPlt(dfPlot,1e-4)
    ggsave(paste0(root_,name_,".png"), p, bg = "transparent")
    return(p)

}

dfPltManhattanHwe = NULL
dfPltAxisManhattanHwe = NULL
dfPltManhattanMaf = NULL
dfPltAxisManhattanMaf = NULL


names_ = c('Only europeans' , 'Known phenotypes' , 'All genotyped individuals')
filters_ = c('NAfr' , 'Geuvadis' , 'Full')
for(idx_ in 1:3){

    print(paste0('sample:',names_[idx_]))
    auxFiltHwe = finalPltDf %>% filter(Sample == filters_[idx_]) %>% distinct(CHR,SNP,P,Sample,BP)
    auxListHwe = manhattanPltDf(auxFiltHwe,names_[idx_])
    dfPltManhattanHwe = rbind(dfPltManhattanHwe , auxListHwe[[1]])
    dfPltAxisManhattanHwe = rbind(dfPltAxisManhattanHwe , auxListHwe[[2]])
    auxListHwe = NULL
    auxFiltHwe = NULL

}

Sys.time()
plotMHwe = manhattanPlt(dfPltManhattanHwe,dfPltAxisManhattanHwe,4,'hwe')
ggsave(paste0(rootSave,'allFramesHwe2',".png"), plotMHwe, bg = "transparent",width=12, height=6, dpi=300)
Sys.time()

dfTotVol = NULL
cutsHwe = c(.05 , 1e-2 , 1e-3 , 1e-4 , 1e-5 , 1e-6 , 1e-7 , 1e-8 )
cutsMaf = c(.001 , .005 , (1:5)/100)
totSnps = finalPltDf %>% distinct(CHR,SNP) %>% group_by(CHR) %>% summarise(nTot=n()) %>% ungroup()
for(cutHwe_ in cutsHwe){
  for(cutMaf_ in cutsMaf){
    print(paste(cutHwe_,cutMaf_))
    dfCuts = merge(
      finalPltDf %>% filter(P >= cutHwe_ , MAF >= cutMaf_) %>% distinct(Sample,CHR,SNP) %>%  group_by(Sample,CHR) %>% summarise(n=n()) %>% ungroup() %>% mutate(hwe = cutHwe_,maf = cutMaf_)
      ,
      totSnps
    )
    dfTotVol = rbind(dfTotVol,dfCuts)
  }
}


dfBars = rbind(
  dfTotVol %>% select(Sample,CHR,n,hwe,maf)
  ,
  dfTotVol %>% select(Sample,CHR,nTot) %>% rename(n=nTot) %>% mutate(hwe = 0,maf=1)
)

plotVol = dfBars %>% ggplot(aes(x = CHR,y=log10(n) , colour = as.factor(hwe) , linetype = as.factor(maf))) + 
geom_line() +
geom_point() +
facet_wrap(~Sample,ncol = 1) +
theme_bw() +
theme( 
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank()
    ) +
labs(x = 'Chromosome',y= expression(Frequency ~ (log[10]) ) , colour = 'HWE',linetype = 'MAF')

ggsave(paste0(rootSave,'combinedFilters0',".png"), plotVol, bg = "transparent",width=10, height=8, dpi=300)

dfVolSum = dfTotVol %>% group_by(Sample,hwe,maf) %>% summarise(n = sum(n) ) %>% ungroup()
dfVolSum$Tot = sum(totSnps$nTot)
dfVolSum$freqRel = dfVolSum$n/dfVolSum$Tot

plotVolMarg = dfVolSum %>% ggplot(aes(x = -log10(hwe),y=100*freqRel , colour = as.factor(maf) , shape = Sample )) + 
facet_wrap(~Sample, ncol = 1) +
geom_line() +
geom_point(size = 3) +
theme_bw() +
theme( 
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank()
    ) +
labs(x = expression( HWE~ (-log[10]) ),y= 'Relative size of set' , colour = 'MAF')+
scale_y_continuous(breaks = 5*c(0:8) , labels = paste0(5*c(0:8),'%'))
ggsave(paste0(rootSave,'combinedFiltersMarg0',".png"), plotVolMarg, bg = "transparent",width=10, height=8, dpi=300)



dfMaf = NULL
cuts = c(.001 , .005 , (1:5)/100)

for(cut_ in cuts){
  print(cut_)
  dfCuts = merge(
    dfPltManhattanMaf %>% filter(MAF <= cut_) %>% group_by(Sample,CHR) %>% summarise(n=n()) %>% ungroup() %>% mutate(cut = cut_)
    ,
    totSnps
  )
  dfMaf = rbind(dfMaf,dfCuts)
}

dfMaf = dfMaf %>% mutate(freq = n/nTot)

knitr::kable(dfMaf)

plotMafLog = dfMaf %>% ggplot(aes(x = CHR,y=log10(n) , colour = as.factor(cut))) + 
geom_point() +
geom_line() + 
facet_wrap(~Sample,ncol = 1) +
theme_bw() +
theme( 
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank()
    ) +
labs(x = 'Chromosome',y= expression(Frequency ~ (log[10]) ) , colour = 'Cutpoint')

plotMafRelativeFr = dfMaf %>% ggplot(aes(x = CHR,y = freq , colour = as.factor(cut))) + 
geom_point() +
geom_line() + 
facet_wrap(~Sample,ncol = 1) +
theme_bw() +
theme( 
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank()
    ) +
labs(x = 'Chromosome',y= "Relative frequency" , colour = 'Cutpoint')

grobLog = ggplotGrob(plotMafLog)
grobLog = gtable_add_cols(grobLog,unit(0,'mm'))
grobRel = ggplotGrob(plotMafRelativeFr)
grobRel = gtable_add_cols(grobRel,unit(0,'mm'))
bindGrobs = rbind(grobLog,grobRel , size = 'first')


ggsave(paste0(rootSave,'allFramesMafx',".png"), bindGrobs, bg = "transparent",width=10, height=8, dpi=300)

